### PR TITLE
mssp/logos

### PR DESCRIPTION
**Pull request by harshshah-tenable**
- Changes for MSSP Logos Endpoint

### CODE REVIEW

File: tenable/io/v3/mssp/logos/api.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
+'''
+Logos
+========
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:3]: equals length should match the length of Logos string.

File: tenable/io/v3/mssp/logos/api.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            name (str):
+                Name of the logo.
+
+        Returns:
+            :obj:`str`:
+                Return the logo_id.
+
+        Example:
+            >>>> with open('/opt/download1.png', 'rb') as fobj:
+                    resp = tio.v3.mssp.logos.add(fobj, 'logo1')
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:46]: this should be 
```
>>> with open('/opt/download1.png', 'rb') as fobj:
...       resp = tio.v3.mssp.logos.add(fobj, 'logo1')
```

File: tenable/io/v3/mssp/logos/api.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        Returns list of logos
+
+        Args:
+
+        Returns:
+            :obj:`list`:
+                Returns all the logo details as a list.
+
+        Example:
+            >>>> tio.v3.mssp.logos.list()
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:62]: replace `>>>>` to `>>>` at all the places

File: tenable/io/v3/mssp/logos/api.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+        Args:
+            logo_id (UUID):
+                The UUID of the logo
+        Returns:
+            :obj:`str`:
+                Return the logo_id.
+
+        Example:
+            >>>> fobj = open('/opt/test_image.png', 'rb')
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:122]: replace `>>>>` to `>>>` at all the places

File: tenable/io/v3/mssp/logos/api.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                 update_resp = tio.v3.mssp.logos.update(
+                     '0cba902a-bd11-4481-bd28-999c88ffe22f',
+                     name='update_name.png',
+                     logo=fobj)
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:126]: this should be like
```
>>> fobj = open('/opt/test_image.png', 'rb')
...      update_resp = tio.v3.mssp.logos.update(
...      '0cba902a-bd11-4481-bd28-999c88ffe22f',
...      name='update_name.png',
...      logo=fobj)
```

File: tenable/io/v3/mssp/logos/api.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+        Args:
+            logo_id (UUID):
+                The UUID of the logo
+        Returns:
+            None
+
+        Example:
+            >>>> tio.v3.mssp.logos.delete(
+                '61a36add-d29b-4a52-bbce-c8215952ede5')
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:148]: this should be 

```
>>> tio.v3.mssp.logos.delete(
...     '61a36add-d29b-4a52-bbce-c8215952ede5')
```

File: tenable/io/v3/mssp/logos/api.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            account_ids (List):
+                list of accoint_ids to assign logos to.
+
+        Returns:
+            None
+
+        Example:
+            >>>> tio.v3.mssp.logos.assign_logos(
+                    'a39f6b74-9b7f-4372-a7ac-a2a4bcb8dbad',
+                    ['0fc4ef49-2649-4c76-bfa7-c181be3adf26'])
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:169]: this should be 
```
>>> tio.v3.mssp.logos.assign_logos(
...    'a39f6b74-9b7f-4372-a7ac-a2a4bcb8dbad',
...    ['0fc4ef49-2649-4c76-bfa7-c181be3adf26'])
```

File: tenable/io/v3/mssp/logos/api.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                A file-like object to write the contents of the logo to.
+        Returns:
+            :obj:`FileObject`:
+                A file-like object containing the contents of the image
+
+        Example:
+            >>>> with open('/opt/f2.png', 'wb') as f:
+                    tio.v3.mssp.logos.download_png(
+                        'a39f6b74-9b7f-4372-a7ac-a2a4bcb8dbad',
+                        fobj=f)
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:200]: Please follow the python consol syntax in example

File: tenable/io/v3/mssp/logos/api.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        Args:
+            logo_id (UUID):
+                The UUID of the logo
+        Returns:
+            :obj:`str`:
+                Return the encoded base64 image contents.
+
+        Example:
+            >>>> resp = tio.v3.mssp.logos.download_base64(
+                    '61a36add-d29b-4a52-bbce-c8215952ede5')
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:226]: Please follow the python consol syntax in example

File: tenable/io/v3/mssp/logos/api.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        Args:
+            logo_id (UUID):
+                The UUID of the logo
+        Returns:
+            :obj:`dict`:
+                Return the logo details.
+
+        Example:
+            >>>> tio.v3.mssp.logos.details(
+                '61a36add-d29b-4a52-bbce-c8215952ede5')
```
Comment (milanaghera-tenable) [inline tenable/io/v3/mssp/logos/api.py:106]: Please follow the python console syntax in the example

**Review by milanaghera-tenable — COMMENTED** [review]
@harshshah-tenable provided some comments on docstring

File: tenable/io/v3/mssp/logos/api.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+            ....    resp = tio.v3.mssp.logos.add(fobj, 'logo1')
+        '''
+        payload = {'files': {'logo': logo, 'name': name}}
+        return self._post(**payload)['id']
+
+    def list(self):
+        '''
+        Returns list of logos
+
+        Args:
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:55]: Do we require "Args:" here?

File: tenable/io/v3/mssp/logos/api.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+                Return the logo_id.
+
+        Example:
+            >>>> with open('/opt/download1.png', 'rb') as fobj:
+            ....    resp = tio.v3.mssp.logos.add(fobj, 'logo1')
+        '''
+        payload = {'files': {'logo': logo, 'name': name}}
+        return self._post(**payload)['id']
+
+    def list(self):
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:51]: Shouldn't this be a search method now? Do we require list method?

File: tenable/io/v3/mssp/logos/api.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+                Returns all the logo details as a list.
+
+        Example:
+            >>>> tio.v3.mssp.logos.list()
+        '''
+        return self._get()['logos']
+
+    def search(self):
+        '''
+        Get the listing of Logos from MSSP portal.
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:68]: Please use this doc string as a reference for search endpoints
https://github.com/dhvanilshah-tenable/pyTenable/pull/12/files#diff-f08a9da0fb05c1c53ae0da456a95945971dfbf2362c0fa30e541363d3728e227R47

File: tenable/io/v3/mssp/logos/api.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+    '''
+    This class contains methods related to Logos API
+    '''
+    _path = 'api/v3/mssp/logos'
+    _conv_json = True
+    _schema = LogoSchema()
+
+    def add(self, logo: Union[BytesIO, BinaryIO], name: str) -> UUID:
+        '''
+        Adds a new logo
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:32]: Please update the doc string to contain more information. For example:
Adds a logo to the Tenable.io MSSP Portal
Ref: https://developer.tenable.com/reference/io-mssp-logos-create

File: tenable/io/v3/mssp/logos/api.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+                An iterator that handles the page management of the requested
+                records.
+        Examples:
+            Todo
+        '''
+        raise NotImplementedError('Search method will be implemented later')
+
+    def details(self, logo_id: UUID) -> Dict:
+        '''
+        Returns Logo details
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:95]: Returns details for the specified logo in the Tenable.io MSSP Portal.

File: tenable/io/v3/mssp/logos/api.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+        Example:
+            >>>> tio.v3.mssp.logos.details(
+            ....    '61a36add-d29b-4a52-bbce-c8215952ede5')
+        '''
+        return self._get(f'{logo_id}')
+
+    def update(self, logo_id: UUID, **kw) -> str:
+        '''
+        Returns Logo details
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:112]: Updates a logo in the Tenable.io MSSP Portal. This update overwrites the existing logo.

File: tenable/io/v3/mssp/logos/api.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+
+        Args:
+            logo_id (UUID):
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:115]: Args should also contain details of the expected kw args. Like name, logo, etc

File: tenable/io/v3/mssp/logos/api.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+        '''
+        payload = {'account_ids': account_ids, 'logo_id': logo_id}
+        self._schema.load(self._schema.dump({
+            'account_ids': account_ids,
+            'logo_id': logo_id
+        }
+        ))
+        self._api.put('api/v3/mssp/accounts/logos', json=payload)
+
+    def download_png(
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/mssp/logos/api.py:179]: Please update the method declaration like this:
https://github.com/dhvanilshah-tenable/pyTenable/pull/12/files#diff-f08a9da0fb05c1c53ae0da456a95945971dfbf2362c0fa30e541363d3728e227R167

**Review by dhvanilshah-tenable — APPROVED** [review]

**Comment by harshshah-tenable** [comment]
@dhvanilshah-tenable @milanaghera-tenable Updated PR as per your comments!

File: tenable/io/v3/mssp/logos/api.py [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+            name (str):
+                Name of the logo.
+
+        Returns:
+            :obj:`str`:
+                Return the logo_id.
+
+        Example:
+            >>> fobj = open('/opt/test_image.png', 'rb')
+            ...     update_resp = tio.v3.mssp.logos.update(
```
Comment (ngaglani-tenable) [inline tenable/io/v3/mssp/logos/api.py]:
```suggestion
            >>> update_resp = tio.v3.mssp.logos.update(
```

File: tenable/io/v3/mssp/logos/api.py (hunk 11 above)
Comment (alpesh-te) [reply tenable/io/v3/mssp/logos/api.py:61]: Is that taken care? @harshshah-tenable @dhvanilshah-tenable

File: tenable/io/v3/mssp/logos/api.py (hunk 17 above)
Comment (alpesh-te) [reply tenable/io/v3/mssp/logos/api.py:121]: Is that taken care? @harshshah-tenable @dhvanilshah-tenable

**Review by alpesh-te — CHANGES_REQUESTED** [review]

File: tenable/io/v3/mssp/logos/api.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+                iterable (based on return_resp flag). Iterator returns all
+                rows in text/csv format for each call with row headers.
+
+        Returns:
+
+            Iterable:
+                The iterable that handles the pagination for the job.
+
+            requests.Response:
+                If ``return_json`` was set to ``True``, then a response
```
Comment (alpesh-te) [inline tenable/io/v3/mssp/logos/api.py]: Update docstring.

File: tests/io/v3/mssp/logos/test_api.py [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+
+    iterator = api.v3.mssp.logos.search(
+        fields=fields, limit=200
+    )
+    assert isinstance(iterator, SearchIterator)
+
+    event_list = []
+    for event in iterator:
+        event_list.append(event)
+    assert len(event_list) == api_response['pagination']['total']
```
Comment (alpesh-te) [inline tests/io/v3/mssp/logos/test_api.py]: Rewrite it to get rid of loop.

**Review by alpesh-te — CHANGES_REQUESTED** [review]

File: tests/io/v3/mssp/logos/test_api.py (hunk 21 above)
Comment (alpesh-te) [reply tests/io/v3/mssp/logos/test_api.py]: Addressed

File: tenable/io/v3/mssp/logos/api.py (hunk 20 above)
Comment (alpesh-te) [reply tenable/io/v3/mssp/logos/api.py]: Addressed

**Review by alpesh-te — APPROVED** [review]
Applied small fixes and looks good to me

**Pull request merged**